woocommerce/woocommerce-ios
# REST API: Integrate Jetpack setup flow after login

Part of #8919 


## Description

This PR wraps up the integration of the Jetpack setup flow after authentication with the custom wpcom login flow. Jetpack setup should completes for both cases when authenticated with site credentials and when authorizing application password through web view. Changes include:
- Networking layer: Updated `JetpackConnectionRemote` to remove `registerJetpackSiteConnection`. We no longer handle this with cookie-nonce programmatically and will switch to using the site connection URL for users to handle the connection manually.
- Yosemite layer: Updated `JetpackConnectionStore` and action to support fetching WPCom account given a WPCom auth token.
- UI layer
  - Updated `AuthenticatedState` to set up `JetpackConnectionStore` with `AlamofireNetwork` when authenticated with an application password. 
  - Updated `AuthenticatedWebViewController` to accept extra credentials for automatic authentication.
  - Updated `JetpackSetupHostingController` to accept credentials to authenticate the connection web view.
  - Updated `JetpackSetupViewModel` to install Jetpack upon receiving 404 error code when fetching Jetpack plugin details. Also updated logic connection URL following pe5sF9-1le-p2#comment-1942.
  - Updated `JetpackConnectionWebViewModel` to detect completion based on the mobile redirect URL or Jetpack plans page.
  - Updated `JetpackSetupCoordinator` to fetch WPCom account after magic link and integrate the Jetpack setup flow after login succeeds.

## Testing instructions

Prerequisite: Ensure that you have access to a self-hosted site with Woo but not Jetpack.
1. Testing presenting Jetpack setup flow when logged in with a magic link:
- Log in to your test site with site credentials.
- Start the Jetpack setup flow by tapping the Jetpack banner on the dashboard screen or the Install Jetpack row in Menu > Settings screen.
- Tap Log In and enter the email address of a WPCom account that can log in with magic link.
- On the password screen, select Log in with magic link.
- Copy the magic link on the received email and paste it in the Safari app of the simulator to open the app with it.
- Notice that a loading screen is displayed and after a while, it is replaced by the Jetpack setup flow.

2. Testing presenting Jetpack setup flow when logged in with password and no 2FA:
- Log in to your test site with site credentials.
- Start the Jetpack setup flow by tapping the Jetpack banner on the dashboard screen or the Install Jetpack row in Menu > Settings screen.
- Tap Log In and log in with the email address of a WPCom account that doesn't have 2FA.
- Notice that the Jetpack setup flow is displayed after the login completes.

3. Testing presenting Jetpack setup flow when logged in with a password with 2FA:
- Log in to your test site with site credentials.
- Start the Jetpack setup flow by tapping the Jetpack banner on the dashboard screen or the Install Jetpack row in Menu > Settings screen.
- Tap Log In and log in with the email address of a WPCom account that has 2FA setup.
- Notice that the Jetpack setup flow is displayed after the login completes.

4. Testing Jetpack setup.
- Log in to your test site with site credentials or application password authorization URL in a web view.
- Start the Jetpack setup flow by tapping the Jetpack banner on the dashboard screen or the Install Jetpack row in Menu > Settings screen.
- Tap Log In and log in with a WPCom account.
- Notice that the Jetpack setup flow is displayed after the login completes.
- Notice that installation and activation are completed successfully.
- Notice that the site connection URL is displayed for the connection step. If you authenticate with site credentials, you shouldn't need to enter site credentials again; otherwise, you will be prompted to log in.
- After the site connection completes, you should be navigated to the account connection page (where you have to tap the Authorize button to complete the setup).
- Dismiss the web view, the interruption error screen should be displayed.
- Tap the Continue setup button to continue.
- Notice that the web page shows the account connection URL again.
- Tap the Authorize button, the web view should then dismiss, and the setup completes.


## Screenshots

No screencast to avoid flashing my email address here.

---
